### PR TITLE
Fix core export provider

### PR DESCRIPTION
**Pull request by oehme**
Since https://github.com/apache/maven/pull/616, the default CoreExportProvider no longer uses the provided CoreExports, but instead tries (and fails) to discover them itself.

This change fixes that by providing our own custom instance of CoreExportProvider. This allows core extension to contribute exported artifacts and exported packages again, like it used to do before the Maven 4.x upgrade.

### CODE REVIEW

**Comment by oehme** [comment]
The underlying reason why the default logic doesn't work is because the [discovery method](https://github.com/apache/maven/blob/fc2d94f0777998000f7b886d45efb88c28fc09cd/maven-core/src/main/java/org/apache/maven/extension/internal/CoreExtensionEntry.java#L83) calls [ClassRealm.loadResourcesFromImport](https://github.com/codehaus-plexus/plexus-classworlds/blob/master/src/main/java/org/codehaus/plexus/classworlds/realm/ClassRealm.java#L526), which does not handle `META-INF` entries from imported realms. It looks up the name `META-INF/maven/extension.xml` and matches that against the import realms. The[ matching logic](https://github.com/codehaus-plexus/plexus-classworlds/blob/b17db1eae685abd72e2f2f0aad5a40c3bc09fd15/src/main/java/org/codehaus/plexus/classworlds/realm/Entry.java#L71) only accounts for classes/resources in packages, but doesn't account for META-INF. So it never never finds the extension.xml entries.

maven-embedder has the [same problem](https://github.com/apache/maven/blob/master/maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java#L594) now.

**Comment by ppalaga** [comment]
I have resterted the ubuntu-18.04 CI job. Not sure why it has shown up as cancelled.

**Comment by gnodet** [comment]
@oehme thx for the detailed explanation.  Would it be possible to add an IT to test this feature ?

**Comment by oehme** [comment]
Yes, I'll add one

**Comment by oehme** [comment]
@gnodet The test infrastructure assumes that every test case is a single project, but for my case I had to install the extension first before using it in the consuming build. I hacked around this by adding the project dir as an additional parameter, but I'd love suggestions for a more elegant solution.

File: integration-tests/src/test/java/org/mvndaemon/mvnd/it/ExtensionWithApiTest.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        TestClientOutput o = new TestClientOutput();
+        client.execute(o, "org.mvndaemon.mvnd.test.extension.with.api:plugin:mojo")
+                .assertSuccess();
+        o.assertContainsMatchingSubsequence("Hello World!");
+    }
+
+    private void installExtension() throws InterruptedException {
+        TestClientOutput o = new TestClientOutput();
+        ((NativeTestClient) client)
+                .execute(o, parameters.multiModuleProjectDirectory().resolve("extension"), List.of("install"))
```
Comment (ppalaga) [inline integration-tests/src/test/java/org/mvndaemon/mvnd/it/ExtensionWithApiTest.java:54]: Wouldn't standard Maven `-f,--file` cli param serve the purpose?
```suggestion
                .execute(o, List.of("install", "-f", "extension"))
```

**Comment by ppalaga** [comment]
> The test infrastructure assumes that every test case is a single project, but for my case I had to install the extension first before using it in the consuming build. I hacked around this by adding the project dir as an additional parameter, but I'd love suggestions for a more elegant solution.

I have left a suggestion inline.

File: integration-tests/src/test/java/org/mvndaemon/mvnd/it/ExtensionWithApiTest.java (hunk 1 above)
Comment (oehme) [reply integration-tests/src/test/java/org/mvndaemon/mvnd/it/ExtensionWithApiTest.java]: Of course :) Done

**Comment by oehme** [comment]
@ppalaga this should be good to go :)

**Review by ppalaga — APPROVED** [review]
I'll let @gnodet re-review and merge.

**Comment by oehme** [comment]
Would it be possible to get an m2 for this? I'd love to run our extensions full CI build against this to make sure we work properly when 1.0 comes out.

**Comment by gnodet** [comment]
> Would it be possible to get an m2 for this? I'd love to run our extensions full CI build against this to make sure we work properly when 1.0 comes out.

Yes, though I'd like to have a `maven 4.0.0-alpha-4` release and then release `mvnd 1.0.0-m2`.

**Comment by ppalaga** [comment]
Here is a fix for GH Actions not publishing the PR's binaries https://github.com/apache/maven-mvnd/pull/774

**Review by gnodet — APPROVED** [review]

**Comment by ppalaga** [comment]
@oehme, BTW if you now send some dummy PR, it will publish the distribution zips for 3 (of 4, excl. Apple M1) supported platforms  under Checks > Artifacts. You could perhaps use those to test on your side?

**Pull request merged**